src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -678,55 +678,201 @@ private static IEnumerable<TSource> TakeWhileIterator<TSource>(IEnumerable<TSour
+                    sourceEnumerator.Dispose();
+                    throw;
                 }
             }
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
         }
 
         public static IEnumerable<TSource> SkipWhile<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            if (source == null) throw Error.ArgumentNull("source");
-            if (predicate == null) throw Error.ArgumentNull("predicate");
-            return SkipWhileIterator<TSource>(source, predicate);
+            Error.CheckSourceNotNull(source);
+            Error.CheckPredicateNotNull(predicate);
+            return new SkipWhileIterator<TSource>(source, predicate);
         }
 
-        private static IEnumerable<TSource> SkipWhileIterator<TSource>(IEnumerable<TSource> source, Func<TSource, bool> predicate)
+        private class SkipWhileIterator<TSource> : IEnumerable<TSource>
         {
-            bool yielding = false;
-            foreach (TSource element in source)
+            private class Enumerator : IEnumerator<TSource>
+            {
+                private readonly IEnumerator<TSource> _source;
+                private readonly Func<TSource, bool> _predicate;
+                private bool _started;
+                public Enumerator(IEnumerable<TSource> source, Func<TSource, bool> predicate)
+                {
+                    _source = source.GetEnumerator();
+                    _predicate = predicate;
+                }
+                public TSource Current
+                {
+                    get { return _source.Current; }
+                }
+                object IEnumerator.Current
+                {
+                    get { return Current; }
+                }
+                public bool MoveNext()
+                {
+                    if (_started)
+                        return _source.MoveNext();
+                    _started = true;
+                    while (_source.MoveNext())
+                        if (!_predicate(_source.Current))
+                            return true;
+                    return false;
+                }
+                public void Reset()
+                {
+                    _source.Reset();
+                    _started = false;
+                }
+                public void Dispose()
+                {
+                    _source.Dispose();
+                }
+            }
+            private readonly IEnumerable<TSource> _source;
+            private readonly Func<TSource, bool> _predicate;
+            public SkipWhileIterator(IEnumerable<TSource> source, Func<TSource, bool> predicate)
+            {
+                _source = source;
+                _predicate = predicate;
+            }
+            public IEnumerator<TSource> GetEnumerator()
+            {
+                return new Enumerator(_source, _predicate);
```

Comment:
Yes. Based on analysis done a while ago, most iterators are used exactly once and saving allocation of an extra object is a statistical win. That resulted in an optimization that reuses the enumerable object as a first iterator instance. We should not give up that optimization without a proof that it is not longer relevant.